FRCs/frc-00zz.md

```diff
@@ -0,0 +1,73 @@
+As the Filecoin network uses keypairs compatible with BLS12-381, it could trivially implement built-in FVM actors for enabling timelock encryption (providing the [FIP for switching to the new drand network](some-link-here) is accepted).
+
+
+## Abstract
+On 2023-03-01, drand launched the first practical Timelock encryption scheme to mainnet after undergoing a security audit and developing libraries and tooling to make it easier.
+Timelock encryption would enable users to do sealed-bid auctions, create more secure implementations of [RANDAO](some-link-here), schedule social media network messages, prevent [MEV](some-link-here), and much, much more!
+The drand timelock scheme uses [identity based encryption](some-link-here) which exploits the pairing capabilities of BLS12-381 to use a future round number as a public key to encrypt messages which will be decryptable only once that future round has been emitted by the drand network by using the round signature as a private key.
+
+We propose a precompile for FVM that would enable users to timelock encrypt arbitrary data and attempt decryption of previously timelock encrypted data. This would enable users to timelock encrypt some part of the data inside their smart contract, while also enabling users to pass ciphertexts directly into smart contracts for decryption there at a later time. 
+
+## Change Motivation
+To enable timelock encryption and decryption using primitives that are already available in the FVM internals.
+
+## Specification
+
+The following interfaces should be fulfilled in FEVM and native actors respectively:
+
+```solidity
+interface Timelock {
+    function encrypt(bytes message, uint64 roundNumber) returns (bytes)
+    function decrypt(bytes message) returns (bytes, bool)
+}
+```
+
+```rust
+trait Timelock {
+    fn encrypt(message: Vec<u8>, roundNumber: u64) -> Result<Vec<u8>>
+    fn decrypt(message: Vec<u8>) -> Result<Vec<u8>>
+}
+```
+
+The `encrypt` functions take raw message bytes and a round number which, along with the public key of the drand network, are passed into the function specified in [the timelock encryption paper](some-link-here) released by the team. In short, the round number is hashed to a point on the G2 extension field of the BLS12-381 curve. That point is multiplied by a new point derived from the message and mapped onto a target group on the Gt extension field, and the message is xor'd with the resulting point on the Gt extension field.
+For larger messages originating outside the system, we recommend using a symmetric cipher such as [AGE](some-link-here) to encrypt the message, and performing timelock encryption on just the symmetric key, in order to save gas fees, as the pairing operation is comparatively expensive.
```

Comment:
```suggestion
For larger messages originating outside the system, we recommend using a symmetric cipher such as [age](https://age-encryption.org/) to encrypt the message, and performing timelock encryption on just the symmetric key, in order to save gas fees, as the pairing operation is comparatively expensive.
```

Is doing on-chain `age` decryption realistic?

Reply:
Or should we consider having a XOF based system doing just XOR?

Reply:
>Is doing on-chain age decryption realistic?
I imagine it's _technically_ possible, but would be very expensive. 
+ cheaper gas fees
+ more flexible if users wish to use other schemes
- we'd need to modify the tlock libraries to support this scenario
- it gives users more scope to do it wrong and try and encrypt huge payloads with tlock

Reply:
I don't think age on-chain has cheaper gas fees! On the contrary

Reply:
So is the message here that users shouldn't timelock their content onchain but only their keys? That could be clearer I think

Reply:
oops yeah dunno why I said it has cheaper gas fees? also the first to were meant to be + and the latter two -

Agree re: messaging, will update